pkg/controller/job/job_controller.go

```diff
@@ -518,6 +518,11 @@ func (jm *Controller) syncJob(key string) (bool, error) {
 		jobFailed = true
 		failureReason = "DeadlineExceeded"
 		failureMessage = "Job was active longer than specified deadline"
+	} else if job.Spec.ActiveDeadlineSeconds != nil {
```

Comment:
Why this is needed? Theoretically when we're creating a new job we should land in the check a few lines above, where we check nil in startTime and accordingly place the key at the edge of its activeDeadlineSeconds. Can you proof your point with a unit test?

Reply:
@soltysh, as I explained in the PR description, this is for when the active deadline seconds is changed while the job is running. 
The bit of code you're referring to only triggers on job creation as that's when start time is nil. 
If you modify the active deadline seconds to be greater than the original, you only get a sync at that original deadline, not the new longer one, so your job will never exceed it's active deadline. 

I couldn't think of a good unit test for this, but I'll have a further think.

Reply:
Right, but job modification is handled already in https://github.com/kubernetes/kubernetes/blob/ca7a906ab793c13f3def2d7ff0357ab88bfecf92/pkg/controller/job/job_controller.go#L338 if ADS changes.

Reply:
Sorry, I see what you've meant, I missed the bit about the workqueue explanation from the PR description. So the problem here is that we are replacing a pre-existing resource, and during name collision we can't queue a new resource with a different (here further in time) delay. The collision is coming from the fact that the queue key is `namespace/name` which will duplicate. So although your approach is good, it will significantly increase the call to the queue, iow. it will try to add to the queue every time ADS is being set. Since this is more of edge case I wonder what would be the best option to handle this.  Maybe we should remove the item from the queue on job deletion. We currently only enqueue the job back.